### PR TITLE
feat(pwsh): cache prompt for repainting

### PR DESCRIPTION
**Pull request by lewis-yeung**
### Prerequisites

- [x] I have read and understood the [contributing guide][CONTRIBUTING.md].
- [x] The commit message follows the [conventional commits][cc] guidelines.
- [ ] Tests for the changes have been added (for bug fixes / features).
- [ ] Docs have been added/updated (for bug fixes / features).

### Description

A follow-up of #5070, making prompt behaviors in PowerShell consistent with other shells.

[CONTRIBUTING.md]: https://github.com/JanDeDobbeleer/oh-my-posh/blob/main/CONTRIBUTING.md
[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

### CODE REVIEW

File: src/prompt/engine.go [hunk 1]
```diff
@@ -500,3 +509,80 @@ func (e *Engine) adjustTrailingDiamondColorOverrides() {
...
+
+func (e *Engine) updatePromptCache(value *promptCache) {
+	cacheJSON, err := json.Marshal(value)
+	if err != nil {
+		return
+	}
+	e.Env.Cache().Set(cache.PROMPTCACHE, string(cacheJSON), 1440)
+}
+
+// New returns a prompt engine initialized with the
```
Comment (JanDeDobbeleer) [inline src/prompt/engine.go:541]: why did you move this one?

File: src/prompt/primary.go [hunk 2]
```diff
@@ -9,71 +9,71 @@ import (
...
-	var cancelNewline, didRender bool
-
 	needsPrimaryRPrompt := e.needsPrimaryRPrompt()
 
-	for i, block := range e.Config.Blocks {
-		// do not print a leading newline when we're at the first row and the prompt is cleared
-		if i == 0 {
-			row, _ := e.Env.CursorPosition()
-			cancelNewline = e.Env.Flags().Cleared || e.Env.Flags().PromptCount == 1 || row == 1
+	switch e.Env.Shell() {
```
Comment (JanDeDobbeleer) [inline src/prompt/primary.go]: no convinced that a switch is the best use-case for this one, why did you do in that direction?

File: src/prompt/tooltip.go [hunk 3]
```diff
@@ -56,6 +43,15 @@ func (e *Engine) Tooltip(tip string) string {
 
 	switch e.Env.Shell() {
 	case shell.PWSH, shell.PWSH5:
+		defer func() {
+			if e.fetchPromptCache() {
```
Comment (JanDeDobbeleer) [inline src/prompt/tooltip.go]: I would rename that `CachePrompt()` as this isn't about fetching.

**Review by JanDeDobbeleer — CHANGES_REQUESTED** [review]
A few comments

File: src/prompt/engine.go (hunk 1 above)
Comment (lewis-yeung) [reply src/prompt/engine.go:541]: Just an advice on the code organization. The function returns a new prompt engine and it's all about the "engine" logic, so there is no need to put it in a standalone source file, IMO. It's okay to revert this if you prefer the original way. :)

Comment (JanDeDobbeleer) [reply src/prompt/engine.go:541]: @lewis-yeung I think we can argue either way. This is fine, I'll allow it 😃

File: src/prompt/primary.go (hunk 2 above)
Comment (lewis-yeung) [reply src/prompt/primary.go]: I'll reorganize the part of logic that adapts to PowerShell only.

File: src/prompt/tooltip.go (hunk 3 above)
Comment (lewis-yeung) [reply src/prompt/tooltip.go]: Not sure if I get your point. Do you mean this is not about "fetching cache from file"? Here it tries to check any existing prompt cache and fill corresponding fields. The return value indicates whether a prompt cache is available.

Comment (lewis-yeung) [reply src/prompt/tooltip.go]: Renamed to `checkPromptCache` now.
